README.md
# dav

[![Build Status](https://travis-ci.org/lambdabaa/dav.png?branch=master)](https://travis-ci.org/lambdabaa/dav)

WebDAV, CalDAV, and CardDAV client for nodejs and the browser.

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

**Table of Contents** _generated with [DocToc](http://doctoc.herokuapp.com/)_

- [API](#api)
  - [accounts](#accounts)
    - [dav.createAccount(options)](#davcreateaccountoptions)
  - [calendars](#calendars)
    - [dav.createCalendarObject(calendar, options)](#davcreatecalendarobjectcalendar-options)
    - [dav.updateCalendarObject(calendarObject, options)](#davupdatecalendarobjectcalendarobject-options)
    - [dav.deleteCalendarObject(calendarObject, options)](#davdeletecalendarobjectcalendarobject-options)
    - [dav.syncCalendar(calendar, options)](#davsynccalendarcalendar-options)
    - [dav.syncCaldavAccount(account, options)](#davsynccaldavaccountaccount-options)
  - [contacts](#contacts)
    - [dav.createCard(addressBook, options)](#davcreatecardaddressbook-options)
    - [dav.updateCard(card, options)](#davupdatecardcard-options)
    - [dav.deleteCard(card, options)](#davdeletecardcard-options)
    - [dav.syncAddressBook(addressBook, options)](#davsyncaddressbookaddressbook-options)
    - [dav.syncCarddavAccount(account, options)](#davsynccarddavaccountaccount-options)
  - [sandbox](#sandbox)
    - [dav.Sandbox()](#davsandbox)
  - [transport](#transport)
    - [dav.transport.Basic(credentials)](#davtransportbasiccredentials)
      - [dav.transport.Basic.send(request, options)](#davtransportbasicsendrequest-options)
    - [dav.transport.OAuth2(credentials)](#davtransportoauth2credentials)
      - [dav.transport.OAuth2.send(request, options)](#davtransportoauth2sendrequest-options)
  - [request](#request)
    - [dav.request.addressBookQuery(options)](#davrequestaddressbookqueryoptions)
    - [dav.request.basic(options)](#davrequestbasicoptions)
    - [dav.request.calendarQuery(options)](#davrequestcalendarqueryoptions)
    - [dav.request.propfind(options)](#davrequestpropfindoptions)
    - [dav.request.syncCollection(options)](#davrequestsynccollectionoptions)
  - [Client](#client)
    - [dav.Client(xhr, options)](#davclientxhr-options)
      - [dav.Client.send(req, options)](#davclientsendreq-options)
  - [etc](#etc)
    - [dav.ns](#davns)
  - [Example Usage](#example-usage)
    - [Using the lower-level webdav request api](#using-the-lower-level-webdav-request-api)
- [Debugging](#debugging)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## API

### accounts

#### dav.createAccount(options)

Perform an initial download of a caldav or carddav account's data. Returns a [Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise) which will be fulfilled with a [dav.Account](https://github.com/gaye/dav/blob/master/lib/model/account.js) object.

```
Options:

  (String) accountType - one of 'caldav' or 'carddav'. Defaults to 'caldav'.
  (Array.<Object>) filters - list of caldav filters to send with request.
  (Boolean) loadCollections - whether or not to load dav collections.
  (Boolean) loadObjects - whether or not to load dav objects.
  (dav.Sandbox) sandbox - optional request sandbox.
  (String) server - some url for server (needn't be base url).
  (String) timezone - VTIMEZONE calendar object.
  (dav.Transport) xhr - request sender.
```

### calendars

#### dav.createCalendarObject(calendar, options)

Create a calendar object on the parameter calendar. Returns a [Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise) which will be fulfilled when the calendar has been created.

```
@param {dav.Calendar} calendar the calendar to put the object on.

Options:

  (String) data - rfc 5545 VCALENDAR object.
  (String) filename - name for the calendar ics file.
  (dav.Sandbox) sandbox - optional request sandbox.
  (dav.Transport) xhr - request sender.
```

#### dav.updateCalendarObject(calendarObject, options)

Persist updates to the parameter calendar object to the server. Returns a [Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise) which will be fulfilled when the calendar has been updated.

```
@param {dav.CalendarObject} calendarObject updated calendar object.

Options:

  (dav.Sandbox) sandbox - optional request sandbox.
  (dav.Transport) xhr - request sender.
```

#### dav.deleteCalendarObject(calendarObject, options)

Delete the parameter calendar object on the server. Returns a [Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise) which will be fulfilled when the calendar has been deleted.

```
@param {dav.CalendarObject} calendarObject target calendar object.

Options:

  (dav.Sandbox) sandbox - optional request sandbox.
  (dav.Transport) xhr - request sender.
```

#### dav.syncCalendar(calendar, options)

Fetch changes from the remote server to the parameter calendar. Returns a [Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise) which will be fulfilled with an updated [dav.Calendar](https://github.com/gaye/dav/blob/master/lib/model/calendar.js) object once sync is complete.

```
@param {dav.Calendar} calendar the calendar to fetch changes for.

Options:

  (Array.<Object>) filters - list of caldav filters to send with request.
  (dav.Sandbox) sandbox - optional request sandbox.
  (String) syncMethod - either 'basic' or 'webdav'. If unspecified, will
      try to do webdav sync and failover to basic sync if rfc 6578 is not
      supported by the server.
  (String) timezone - VTIMEZONE calendar object.
  (dav.Transport) xhr - request sender.
```

#### dav.syncCaldavAccount(account, options)

Fetch changes from the remote server to the account's calendars. Returns a [Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise) which will be fulfilled with an updated [dav.Account](https://github.com/gaye/dav/blob/master/lib/model/account.js) object once sync is complete.

```
@param {dav.Account} account the calendar account to sync.

Options:

  (dav.Sandbox) sandbox - optional request sandbox.
  (dav.Transport) xhr - request sender.
```

### contacts

#### dav.createCard(addressBook, options)

Create a vcard object on the parameter address book. Returns a [Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise) which will be fulfilled when the vcard has been created.

```
@param {dav.AddressBook} addressBook the address book to put the object on.

Options:

  (String) data - VCARD object.
  (String) filename - name for the vcard vcf file.
  (dav.Sandbox) sandbox - optional request sandbox.
  (dav.Transport) xhr - request sender.
```

#### dav.updateCard(card, options)

Persist updates to the parameter vcard object to the server. Returns a [Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise) which will be fulfilled when the vcard has been updated.

```
@param {dav.VCard} card updated vcard object.

Options:

  (dav.Sandbox) sandbox - optional request sandbox.
  (dav.Transport) xhr - request sender.
```

#### dav.deleteCard(card, options)

Delete the parameter vcard object on the server. Returns a [Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise) which will be fulfilled when the vcard has been deleted.

```
@param {dav.VCard} card target vcard object.

Options:

  (dav.Sandbox) sandbox - optional request sandbox.
  (dav.Transport) xhr - request sender.
```

#### dav.syncAddressBook(addressBook, options)

Fetch changes from the remote server to the parameter address books. Returns a [Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise) which will be fulfilled with an updated [dav.AddressBook](https://github.com/gaye/dav/blob/master/lib/model/address_book.js) object once sync is complete.

```
@param {dav.AddressBook} addressBook the address book to fetch changes for.

Options:

  (dav.Sandbox) sandbox - optional request sandbox.
  (String) syncMethod - either 'basic' or 'webdav'. If unspecified, will
      try to do webdav sync and failover to basic sync if rfc 6578 is not
      supported by the server.
  (dav.Transport) xhr - request sender.
```

#### dav.syncCarddavAccount(account, options)

Fetch changes from the remote server to the account's address books. Returns a [Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise) which will be fulfilled with an updated [dav.Account](https://github.com/gaye/dav/blob/master/lib/model/account.js) object once sync is complete.

```
@param {dav.Account} account the address book account to sync.

Options:

  (dav.Sandbox) sandbox - optional request sandbox.
  (dav.Transport) xhr - request sender.
```

### sandbox

#### dav.Sandbox()

Create a request sandbox. There is also a deprecated interface
`dav.createSandbox()`. Add requests to the sandbox like so:

```js
var sandbox = new dav.Sandbox();
// sandbox instanceof Sandbox
dav
  .createAccount({
    username: 'Yoshi',
    password: 'babybowsersoscaryomg',
    server: 'https://caldav.yoshisstory.com',
    sandbox: sandbox, // <- Insert sandbox here!
  })
  .then(function (calendars) {
    // etc, etc.
  });
```

And abort sandboxed requests as a group with `sandbox.abort()`.

### transport

#### dav.transport.Basic(credentials)

Create a new `dav.transport.Basic` object. This sends dav requests using http basic authentication.

```
@param {dav.Credentials} credentials user authorization.
```

##### dav.transport.Basic.send(request, options)

```
@param {dav.Request} request object with request info.
@return {Promise} a promise that will be resolved with an xhr request after its readyState is 4 or the result of applying an optional request `transformResponse` function to the xhr object after its readyState is 4.

Options:

  (dav.Sandbox) sandbox - optional request sandbox.
```

#### dav.transport.OAuth2(credentials)

Create a new `dav.transport.OAuth2` object. This sends dav requests authorized via rfc 6749 oauth2.

```
@param {dav.Credentials} credentials user authorization.
```

##### dav.transport.OAuth2.send(request, options)

```
@param {dav.Request} request object with request info.
@return {Promise} a promise that will be resolved with an xhr request after its readyState is 4 or the result of applying an optional request `transformResponse` function to the xhr object after its readyState is 4.

Options:

  (dav.Sandbox) sandbox - optional request sandbox.
```

### request

#### dav.request.addressBookQuery(options)

```
Options:

  (String) depth - optional value for Depth header.
  (Array.<Object>) props - list of props to request.
```

#### dav.request.basic(options)

```
Options:

  (String) data - put request body.
  (String) method - http method.
  (String) etag - cached calendar object etag.
```

#### dav.request.calendarQuery(options)

```
Options:

  (String) depth - optional value for Depth header.
  (Array.<Object>) filters - list of filters to send with request.
  (Array.<Object>) props - list of props to request.

HISTORY.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

**Table of Contents** _generated with [DocToc](http://doctoc.herokuapp.com/)_

- [1.7.8](#178)
- [1.7.7](#177)
- [1.7.6](#176)
- [1.7.5](#175)
- [1.7.4](#174)
- [1.7.3](#173)
- [1.7.2](#172)
- [1.7.0](#170)
- [1.6.5](#165)
- [1.6.4](#164)
- [1.6.3](#163)
- [1.6.2](#162)
- [1.6.1](#161)
- [1.6.0](#160)
- [1.5.5](#155)
- [1.5.4](#154)
- [1.5.3](#153)
- [1.5.2](#152)
- [1.5.1](#151)
- [1.5.0](#150)
- [1.4.1](#141)
- [1.4.0](#140)
- [1.3.0](#130)
- [1.2.0](#120)
- [1.1.2](#112)
- [1.1.1](#111)
- [1.1.0](#110)
- [1.0.4](#104)
- [1.0.3](#103)
- [1.0.2](#102)
- [1.0.1](#101)
- [1.0.0](#100)
- [0.11.0](#0110)
- [0.10.1](#0101)
- [0.10.0](#0100)
- [0.9.3](#093)
- [0.9.2](#092)
- [0.9.1](#091)
- [0.9.0](#090)
- [0.8.0](#080)
- [0.7.1](#071)
- [0.7.0](#070)
- [0.6.0](#060)
- [0.5.0](#050)
- [0.4.0](#040)
- [0.3.1](#031)
- [0.3.0](#030)
- [0.2.0](#020)
- [0.1.0](#010)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

### 1.7.8

- `listCalendars` passes through all calendars that contain iCalendar
  objects and not just `VEVENT`s.

### 1.7.7

- Parser processes href and currentUserPrincipal elements

### 1.7.6

- Bundle `regeneratorRuntime` and polyfills for `Array.prototype.find`
  and `Object.assign` so that consumers don't need babel polyfill.

### 1.7.5

- Trimmed the binary size to 128k minified
- Remove babel polyfill
- Remove handlebars

### 1.7.4

- Precompile handlebars templates
- Use custom debug module

### 1.7.3

- Google CardDAV wasn't working due to a few issues like our url
  comparison, Google requiring at least one filter for an addressbook
  query, and Google sometimes returning propstats with statuses but no
  props.

### 1.7.2

- Fix outstanding nodejs compatibility issues
- Expose dav.version

### 1.7.0

- Remove lodash, json-stringify-safe, don't expose jsonify

### 1.6.5

- Add brfs as a dependency so that downstream consumers can browserify us

### 1.6.4

- Workaround https://github.com/substack/brfs/issues/39

### 1.6.3

- Convert to use new es6/7 features and transpile with babel

### 1.6.2

- Export debug library under dav ns

### 1.6.1

- Don't bundle xmlhttprequest polyfill in browser binary... again

### 1.6.0

- Add #syncCaldavAccount and #syncCarddavAccount to the public api
- Expose dav.jsonify and dav.ns
- Small correctness fix to error case in basic calendar sync

### 1.5.5

- Bundle XMLHttpRequestWrapper polyfill for environments where it's not available

### 1.5.4

- Fix browser globals

### 1.5.3

- Don't use window in web workers

### 1.5.2

- Use xmldom in the browser since it's missing from web workers

### 1.5.1

- Expose dav browserify configuration to npm consumers

### 1.5.0

- Decouple requests from the urls they get sent to

### 1.4.1

- Add missing use strict statement to lib/index.js

### 1.4.0

- New sandbox interface

### 1.3.0

- Expose dav.Model, dav.Request, dav.Transport

### 1.2.0

- Implement client#send

### 1.1.2

- Trick browserify into not bundling node shims for web apis

### 1.1.1

- %s/toString/jsonify/ for models

### 1.1.0

- Support for rfc 6352 carddav

### 1.0.4

- Implement #toString on models

### 1.0.3

- Internal DELETE, PUT request refactor

### 1.0.2

- davinci -> dav

### 1.0.1

- Fix issues with browserify build

### 1.0.0

- Update interfaces for pluggable transports, expose transport layer
- Support for oauth2 authentication
- Clean up internal multistatus parser

### 0.11.0

- Support for rfc 6578 webdav sync

### 0.10.1

- Set request depth to 0 in the "getctag" propfind issued during sync

### 0.10.0

- Implement time-range filters for calendar queries

### 0.9.3

- Remove dependencies on ical.js and underscore

### 0.9.2

- Fix npm package
- Change npm name to davincijs

### 0.9.1

- remove nodejs polyfills for DOMParser and XMLHttpRequestWrapper from build output
- generate minified binaries

### 0.9.0

- Implement davinci.Client interface
- Add transport layer to decouple request details and sending

### 0.8.0

- Expose low-level request methods through davinci.request
- Add hook to requests to override transformResponse

### 0.7.1

- Expose the underlying, xml parsed dav responses on davinci.Calendar and davinci.CalendarObject models.

### 0.7.0

- Support providing timezone option to #createAccount and #syncCalendar

### 0.6.0

- #syncCalendar added to public api
- The promise returned from #createAccount now resolves with a davinci.Account object instead of an array of davinci.Calendar objects.

### 0.5.0

- #deleteCalendarObject added to public api

### 0.4.0

- #updateCalendarObject added to public api
- Internal api refactoring to expose Request objects

### 0.3.1

- Patch bug in build due to bug in brfs.

### 0.3.0

- #createCalendarObject modified to support sandboxing.

### 0.2.0

- #createCalendarObject added to public api

### 0.1.0

- #createAccount added to public api
- #createSandbox added to public api


index.d.ts
// Type definitions for dav 1.7
// Project: https://github.com/lambdabaa/dav/, https://github.com/gaye/dav
// Definitions by: ToastHawaii <https://github.com/ToastHawaii>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
// TypeScript Version: 2.2

export let version: string;

/**
 * Perform an initial download of a caldav or carddav account's data.
 * @param options
 * @returns a Promise which will be fulfilled with a dav.Account object.
 */
export function createAccount(options: CreateAccountOptions): Promise<Account>;

export interface CreateAccountOptions {
  /**
   * one of 'caldav' or 'carddav'. Defaults to 'caldav'.
   */
  accountType?: 'caldav' | 'carddav';

  /**
   * list of caldav filters to send with request.
   */
  filters?: object[];

  /**
   *  whether or not to load dav collections.
   */
  loadCollections?: boolean;

  /**
   * whether or not to load dav objects.
   */
  loadObjects?: boolean;

  /**
   * request sandox.
   */
  sandbox?: Sandbox | {};

  /**
   *  some url for server (needn't be base url).
   */
  server: string;

  /**
   * VTIMEZONE calendar object.
   */
  timezone?: string;

  /**
   * request sender.
   */

  xhr?: transport.Transport;
}

/**
 * Create a calendar object on the parameter calendar.
 * @param calendar the calendar to put the object on.
 * @param options
 * @returns a Promise which will be fulfilled when the calendar has been created.
 */
export function createCalendarObject(
  calendar: Calendar,
  options: CreateCalendarObjectOptions
): Promise<CalendarObject>;

export interface CreateCalendarObjectOptions {
  /**
   * rfc 5545 VCALENDAR object.
   */
  data: string;

  /**
   * name for the calendar ics file.
   */
  filename: string;

  /**
   * request sandbox.
   */
  sandbox?: Sandbox;

  /**
   * request sender.
   */
  xhr?: transport.Transport;
}

/**
 * Persist updates to the parameter calendar object to the server.
 * @param calendarObject updated calendar object.
 * @param options
 * @returns a Promise which will be fulfilled when the calendar has been updated.
 */
export function updateCalendarObject(
  calendarObject: CalendarObject,
  options: UpdateCalendarObjectOptions
): Promise<CalendarObject>;

export interface UpdateCalendarObjectOptions {
  /**
   * request sandbox.
   */
  sandbox?: Sandbox;

  /**
   * request sender.
   */
  xhr?: transport.Transport;
}

/**
 * Delete the parameter calendar object on the server.
 * @param calendarObject target calendar object.
 * @param options
 * @returns a Promise which will be fulfilled when the calendar has been deleted.
 */
export function deleteCalendarObject(
  calendarObject: CalendarObject,
  options: DeleteCalendarObjectOptions
): Promise<CalendarObject>;

export interface DeleteCalendarObjectOptions {
  /**
   * request sandbox.
   */
  sandbox?: Sandbox;

  /**
   * request sender.
   */
  xhr?: transport.Transport;
}

/**
 * Fetch changes from the remote server to the parameter calendar.
 * @param calendar the calendar to fetch changes for.
 * @param options
 * @returns a Promise which will be fulfilled with an updated dav.Calendar object once sync is complete.
 */
export function syncCalendar(
  calendar: Calendar,
  options: SyncCalendarOptions
): Promise<Calendar>;

export interface SyncCalendarOptions {
  /**
   * list of caldav filters to send with request.
   */
  filters?: object[];

  /**
   * request sandbox.
   */
  sandbox?: Sandbox;

  /**
   * either 'basic' or 'webdav'. If unspecified, will try to do webdav sync
   * and failover to basic sync if rfc 6578 is not supported by the server.
   */
  syncMethod?: 'basic' | 'webdav';

  /**
   * VTIMEZONE calendar object.
   */
  timezone?: string;

  /**
   * request sender.
   */
  xhr?: transport.Transport;
}

/**
 * Fetch changes from the remote server to the account's calendars.
 * @param account the calendar account to sync.
 * @param options
 * @returns a Promise which will be fulfilled with an updated dav.Account object once sync is complete.
 */
export function syncCaldavAccount(
  account: Account,
  options: SyncCaldavAccountOptions
): Promise<Account>;

export interface SyncCaldavAccountOptions {
  /**
   * list of caldav filters to send with request.
   */
  filters?: object[];

  /**
   * request sandbox.
   */
  sandbox?: Sandbox;

  /**
   * either 'basic' or 'webdav'. If unspecified, will try to do webdav sync
   * and failover to basic sync if rfc 6578 is not supported by the server.
   */
  syncMethod?: 'basic' | 'webdav';

  /**
   * VTIMEZONE calendar object.
   */
  timezone?: string;

  /**
   * request sender.
   */
  xhr?: transport.Transport;
}

/**
 * Create a vcard object on the parameter address book.
 * @param addressBook the address book to put the object on.
 * @param options
 * @returns a Promise which will be fulfilled when the vcard has been created.
 */
export function createCard(
  addressBook: AddressBook,
  options: CreateCardOptions
): Promise<AddressBook>;

export interface CreateCardOptions {
  /**
   * VCARD object.
   */
  data: string;
  /**
   * name for the vcard vcf file.
   */
  filename: string;

  /**
   * request sandbox.
   */
  sandbox?: Sandbox;
  /**
   * request sender.
   */
  xhr?: transport.Transport;
}

/**
 *     Persist updates to the parameter vcard object to the server.
 * @param  card updated vcard object.
 * @param options
 * @returns a Promise which will be fulfilled when the vcard has been updated.
 */
export function updateCard(
  card: VCard,
  options: UpdateCardOptions
): Promise<VCard>;

export interface UpdateCardOptions {
  /**
   * request sandbox.
   */
  sandbox?: Sandbox;
  /**
   * request sender.
   */
  xhr?: transport.Transport;
}

/**
 * Delete the parameter vcard object on the server.
 * @param card target vcard object.
 * @param options
 * @returns a Promise which will be fulfilled when the vcard has been deleted.
 */
export function deleteCard(
  card: VCard,
  options: DeleteCardOptions
): Promise<VCard>;

export interface DeleteCardOptions {
  /**
   * request sandbox.
   */
  sandbox?: Sandbox;
  /**
   * request sender.
   */
  xhr?: transport.Transport;
}

/**
 * Fetch changes from the remote server to the parameter address books.
 * @param addressBook the address book to fetch changes for.
 * @param options
 * @returns a Promise which will be fulfilled with an updated AddressBook object once sync is complete.
 */
export function syncAddressBook(
  addressBook: AddressBook,
  options: SyncAddressBookOptions
): Promise<AddressBook>;

export interface SyncAddressBookOptions {
  /**
   * request sandbox.
   */
  sandbox?: Sandbox;
  /**
   * either 'basic' or 'webdav'.If unspecified, will try to do webdav sync
   * and failover to basic sync if rfc 6578 is not supported by the server.
   */
  syncMethod?: 'basic' | 'webdav';

  /**
   * request sender.
   */
  xhr?: transport.Transport;
}

/**
 * Fetch changes from the remote server to the account's address books.
 * @param  account the address book account to sync.
 * @param options
 * @returns a Promise which will be fulfilled with an updated Account object once sync is complete.
 */
export function syncCarddavAccount(
  account: Account,
  options: SyncCarddavAccountOptions
): Promise<Account>;

export interface SyncCarddavAccountOptions {
  /**
   * list of caldav filters to send with request.
   */
  filters?: object[];

  /**
   * request sandbox.
   */
  sandbox?: Sandbox;

  /**
   * either 'basic' or 'webdav'. If unspecified, will try to do webdav sync
   * and failover to basic sync if rfc 6578 is not supported by the server.
   */
  syncMethod?: 'basic' | 'webdav';

  /**
   * VTIMEZONE calendar object.
   */
  timezone?: string;

  /**
   * request sender.
   */
  xhr?: transport.Transport;
}

/**
 * Create a request sandbox.
 */
export class Sandbox {
  constructor();

  requestList: any[];

  add(request: any): void;

  /**
   * abort sandboxed requests as a group.
   */
  abort(): void;
}

/**
 * @deprecated
 */
export function createSandbox(): Sandbox;

export namespace transport {
  class Transport {
    /**
     * @param credentials user authorization.
     */
    constructor(credentials: Credentials);

    send(
      request: Request,
      url: string,
      options?: TransportOptions
    ): Promise<any>;
  }

  interface TransportOptions {
    /**
     *  request sandbox.
     */
    sandbox?: Sandbox;

    retry?: boolean;
  }

  class Basic extends Transport {
    /**
     * Create a new Basic object. This sends dav requests using http basic authentication.
     * @param credentials user authorization.
     */
    constructor(credentials: Credentials);

    /**
     *
     * @param request object with request info.
     * @param url
     * @param options
     * @return a promise that will be resolved with an xhr request after
     * its readyState is 4 or the result of applying an optional request
     * `transformResponse` function to the xhr object after its readyState
     * is 4.
     */
    send(
      request: Request,
      url: string,
      options?: TransportOptions
    ): Promise<any>;
  }

  /**
   * Create a new OAuth2 object.This sends dav requests authorized via rfc 6749 oauth2.
   * @param credentials user authorization.
   */
  class OAuth2 extends Transport {
    constructor(credentials: Credentials);

    /**
     *
     * @param request object with request info.
     * @param url
     * @param options
     * @return a promise that will be resolved with an xhr request after
     * its readyState is 4 or the result of applying an optional request
     * `transformResponse` function to the xhr object after its readyState
     * is 4.
     */
    send(
      request: Request,
      url: string,
      options?: TransportOptions
    ): Promise<any>;
  }
}

export namespace request {
  /**
   *
   * @param options
   * @returns
   */
  function addressBookQuery(options: AddressBookQueryOptions): string;

  interface AddressBookQueryOptions {
    /**
     * value for Depth header.
     */
    depth?: string;

    /**
     * list of props to request.
     */
    props: object[];
  }

  /**
   *
   * @param options
   * @returns
   */
  function basic(options: BasicOptions): Request;

  interface BasicOptions {
    /**
     * put request body.
     */
    data: string;

    /**
     * http method.
     */
    method: string;

    /**
     * cached calendar object etag.
     */
    etag: string;
  }

  /**
   *
   * @param options
   * @returns
   */
  function calendarQuery(options: CalendarQueryOptions): string;

  interface CalendarQueryOptions {
    /**
     * value for Depth header.
     */
    depth?: string;

    /**
     * list of filters to send with request.
     */
    filters: object[];

    /**
     * list of props to request.
     */
    props: object[];

    /**
     * VTIMEZONE calendar object.
     */
    timezone: string;
  }

  /**
   *
   * @param options
   * @returns
   */
  function propfind(options: PropfindOptions): string;

  interface PropfindOptions {
    /**
     *  value for Depth header.
     */
    depth?: string;

    /**
     * list of props to request.
     */
    props: object[];
  }

  /**
   *
   * @param options
   * @returns
   */
  function syncCollection(options: SyncCollectionOptions): string;

  interface SyncCollectionOptions {
    /**
     * option value for Depth header.
     */
    depth?: string;

    /**
     * list of props to request.
     */
    props: object[];

    /**
     * indicates scope of the sync report request.
     */
    syncLevel: number;

    /**
     * synchronization token provided by the server.
     */
    syncToken: string;
  }
}

export class Client {
  /**
   * Create a new Client object. The client interface allows consumers to set
   * their credentials and transport once and then make authorized requests
   * without passing them to each request. Each of the other, public API
   * methods should be available on Client objects.
   * @param xhr request sender.
   * @param options
   */
  constructor(xhr: transport.Transport, options?: ClientOptions);

  /**
   * Send a request using this client's transport (and perhaps baseUrl).
   * @param  req dav request.
   * @param options
   * @return a promise that will be resolved with an xhr request after
   * its readyState is 4 or the result of applying an optional request
   * `transformResponse` function to the xhr object after its readyState
   * is 4.
   */
  send(req: Request, uri: string, options?: ClientSendOptions): Promise<any>;

  /**
   * Perform an initial download of a caldav or carddav account's data.
   * @param options
   * @returns a Promise which will be fulfilled with a dav.Account object.
   */
  createAccount(options?: CreateAccountOptions): Promise<Account>;

  /**
   * Create a calendar object on the parameter calendar.
   * @param calendar the calendar to put the object on.
   * @param options
   * @returns a Promise which will be fulfilled when the calendar has been created.
   */
  createCalendarObject(
    calendar: Calendar,
    options?: CreateCalendarObjectOptions
  ): Promise<CalendarObject>;

  /**
   * Persist updates to the parameter calendar object to the server.
   * @param calendarObject updated calendar object.
   * @param options
   * @returns a Promise which will be fulfilled when the calendar has been updated.
   */
  updateCalendarObject(
    calendarObject: CalendarObject,
    options?: UpdateCalendarObjectOptions
  ): Promise<CalendarObject>;

  /**
   * Delete the parameter calendar object on the server.
   * @param calendarObject target calendar object.
   * @param options
   * @returns a Promise which will be fulfilled when the calendar has been deleted.
   */
  deleteCalendarObject(
    calendarObject: CalendarObject,
    options?: DeleteCalendarObjectOptions
  ): Promise<CalendarObject>;

  /**
   * Fetch changes from the remote server to the parameter calendar.
   * @param calendar the calendar to fetch changes for.
   * @param options
   * @returns a Promise which will be fulfilled with an updated dav.Calendar object once sync is complete.
   */
  syncCalendar(
    calendar: Calendar,
    options?: SyncCalendarOptions
  ): Promise<Calendar>;

  /**
   * Fetch changes from the remote server to the account's calendars.
   * @param account the calendar account to sync.
   * @param options
   * @returns a Promise which will be fulfilled with an updated dav.Account object once sync is complete.
   */
  syncCaldavAccount(
    account: Account,
    options?: SyncCaldavAccountOptions
  ): Promise<Account>;

  /**
   * Create a vcard object on the parameter address book.
   * @param addressBook the address book to put the object on.
   * @param options
   * @returns a Promise which will be fulfilled when the vcard has been created.
   */
  createCard(
    addressBook: AddressBook,
    options?: CreateCardOptions
  ): Promise<AddressBook>;

  /**
   * Persist updates to the parameter vcard object to the server.
   * @param  card updated vcard object.
   * @param options
   * @returns a Promise which will be fulfilled when the vcard has been updated.
   */
  updateCard(card: VCard, options?: UpdateCardOptions): Promise<VCard>;

  /**
   *
   * Delete the parameter vcard object on the server.
   * @param card target vcard object.
   * @param options
   * @returns a Promise which will be fulfilled when the vcard has been deleted.
   */
  deleteCard(card: VCard, options?: DeleteCardOptions): Promise<VCard>;

  /**
   * Fetch changes from the remote server to the parameter address books.
   * @param addressBook the address book to fetch changes for.
   * @param options
   * @returns a Promise which will be fulfilled with an updated AddressBook object once sync is complete.
   */
  syncAddressBook(
    addressBook: AddressBook,
    options?: SyncAddressBookOptions
  ): Promise<AddressBook>;

  /**
   * Fetch changes from the remote server to the account's address books.
   * @param  account the address book account to sync.
   * @param options
   * @returns a Promise which will be fulfilled with an updated Account object once sync is complete.
   */
  syncCarddavAccount(
    account: Account,
    options?: SyncCarddavAccountOptions
  ): Promise<Account>;
}

export interface ClientOptions {
  /**
   * root url to resolve relative request urls with.
   */
  baseUrl: string;
}

export interface ClientSendOptions {
  /**
   * request sandbox.
   */
  sandbox?: Sandbox;
  /**
   * relative url for request.
   */
  url?: string;
}

export type Partial<T> = {
  [P in keyof T]?: T[P];
};

export class Account {
  constructor(options?: AccountOptions);
  server: string;
  credentials: Credentials;
  rootUrl: string;
  principalUrl: string;
  homeUrl: string;
  calendars: Calendar[];
  addressBooks: AddressBook[];
}

export type AccountOptions = Partial<Account>;

export class Credentials {
  constructor(options?: CredentialsOptions);

  /**
   * username (perhaps email) for calendar user.
   */
  username: string;

  /**
   * plaintext password for calendar user.
   */
  password: string;

  /**
   * oauth client id.
   */
  clientId: string;

  /**
   * oauth client secret.
   */
  clientSecret: string;

  /**
   * oauth code.
   */
  authorizationCode: string;

  /**
   * oauth redirect url.
   */
  redirectUrl: string;

  /**
   * oauth token url.
   */
  tokenUrl: string;

  /**
   * oauth access token.
   */
  accessToken: string;

  /**
   * oauth refresh token.
   */
  refreshToken: string;

  /**
   * unix time for access token expiration.
   */
  expiration: number;
}
export type CredentialsOptions = Partial<Credentials>;

export class DAVCollection<T> {
  constructor(options: DAVCollectionOptions<T>);
  data: string;
  objects: T[];
  account: Account;
  ctag: string;
  description: string;
  displayName: string;
  reports: string[];
  resourcetype: string;
  syncToken: string;
  url: string;
}
export type DAVCollectionOptions<T> = Partial<DAVCollection<T>>;

export class AddressBook extends DAVCollection<VCard> {
  constructor(options?: AddressBookOptions);
}
export type AddressBookOptions = Partial<AddressBook>;

export class Calendar extends DAVCollection<CalendarObject> {
  constructor(options?: CalendarOptions);

  components: string[];
  timezone: string;
}
export type CalendarOptions = Partial<Calendar>;

export class DAVObject {
  constructor(options: DAVObjectOptions);
  data: string;
  etag: string;
  url: string;
  status?: string; // equals 'cancelled' if object is cancelled on remote
}
export type DAVObjectOptions = Partial<DAVObject>;

export class CalendarObject extends DAVObject {
  constructor(options?: CalendarObjectOptions);
  calendar: Calendar;
  calendarData: string;
}
export type CalendarObjectOptions = Partial<CalendarObject>;

export class VCard extends DAVObject {
  constructor(options?: VCardOptions);
  addressBook: AddressBook;
  addressData: string;
}

export type VCardOptions = Partial<VCard>;

export class Request {
  constructor(options?: RequestOptions);
  method: string;
  requestData?: string;
  transformRequest?: (xhr: any) => any;
  transformResponse?: (xhr: any) => any;
  onerror?: (error: Error) => any;
}

export type RequestOptions = Partial<Request>;

export namespace debug {
  let enabled: boolean;
}

export namespace ns {
  const CALENDAR_SERVER = 'http://calendarserver.org/ns/';
  const CALDAV_APPLE = 'http://apple.com/ns/ical/';
  const CALDAV = 'urn:ietf:params:xml:ns:caldav';
  const CARDDAV = 'urn:ietf:params:xml:ns:carddav';
  const DAV = 'DAV:';
}
